ext/src/zpages/tracez_data_aggregator.cc

```diff
@@ -0,0 +1,110 @@
+{
+  AggregateSpans();
+  return aggregated_data_;
+}
+
+LatencyBoundaryName TracezDataAggregator::GetLatencyBoundary(opentelemetry::sdk::trace::Recordable* recordable)
+{
+  auto recordable_duration = recordable->GetDuration();
+  for(unsigned int boundary = 0; boundary < kLatencyBoundaries.size()-1; boundary++)
+  {
+    if(kLatencyBoundaries[boundary] <= recordable_duration && recordable_duration < kLatencyBoundaries[boundary+1])return (LatencyBoundaryName)boundary;
+  }
+  return LatencyBoundaryName::k100SecondToMax;
+}
+
+void TracezDataAggregator::AggregateStatusOKSpans(std::unique_ptr<opentelemetry::sdk::trace::Recordable>& ok_span)
+{
+  LatencyBoundaryName boundary_name = GetLatencyBoundary(ok_span.get());
+  std::string span_name = ok_span.get()->GetName().data();
+  
+  //If the sample span storage is at capacity, remove the span that was inserted earliest and free up memory
+  if(aggregated_data_[span_name].get()->latency_sample_spans_[boundary_name].size() == kMaxNumberOfSampleSpans)
+  {
+    aggregated_data_[span_name].get()->latency_sample_spans_[boundary_name].front().reset();
+    aggregated_data_[span_name].get()->latency_sample_spans_[boundary_name].pop_front();
+  }
+  aggregated_data_[span_name].get()->latency_sample_spans_[boundary_name].push_back(std::move(ok_span));
+  aggregated_data_[span_name].get()->span_count_per_latency_bucket_[boundary_name]++;
+}
+
+void TracezDataAggregator::AggregateStatusErrorSpans(std::unique_ptr<opentelemetry::sdk::trace::Recordable>& error_span)
+{
+  std::string span_name = error_span.get()->GetName().data();
+  
+  //If the error samples storage is at capacity, remove the span that was inserted earliest and free up memory
+  if(aggregated_data_[span_name].get()->error_sample_spans_.size() == kMaxNumberOfSampleSpans)
+  {
+    aggregated_data_[span_name].get()->error_sample_spans_.front().reset();
+    aggregated_data_[span_name].get()->error_sample_spans_.pop_front();
+  }
+  aggregated_data_[span_name].get()->error_sample_spans_.push_back(std::move(error_span));
+  aggregated_data_[span_name].get()->error_spans_++;
+}
+
+void TracezDataAggregator::AggregateCompletedSpans()
+{
+  auto completed_spans = tracez_span_processor_->GetCompletedSpans();
+  
+  for(auto& span: completed_spans)
+  {
+    std::string span_name = span.get()->GetName().data();
+    
+    if(aggregated_data_.find(span_name) == aggregated_data_.end())
+    aggregated_data_[span_name] = std::unique_ptr<AggregatedInformation>(new AggregatedInformation);
+    
+    //running spans are calculated from scratch later
+    aggregated_data_[span_name].get() -> running_spans_ = 0;
+    
+    if(span.get()->GetStatus() == opentelemetry::trace::CanonicalCode::OK)AggregateStatusOKSpans(span);
+    else AggregateStatusErrorSpans(span);
+  }
+}
+
+void TracezDataAggregator::AggregateRunningSpans()
+{
+  auto running_spans = tracez_span_processor_->GetRunningSpans();
+  std::unordered_set<std::string> cache;
```

Comment:
How the cache is used?

Reply:
Good question, I wanted to explain this during the standup but here is the overview. cache is used to keep track of the first time that I see a running span. Every time that function is called all it assumes that all running spans for each span name is 0.
Hence whenever I see a span name for the first time I put it into the cache and set it's running span value to 0. After the first time I increment it normally. I can elaborate on this with examples in a meeting.